### DISCUSSION TITLE
Docker Compose not starting webserver

### DISCUSSION THREAD

**Discussion by DanielMorales9** [apache/airflow#19140] [General]
### Apache Airflow version

2.2.0 (latest released)

### Operating System

OSX

### Versions of Apache Airflow Providers

none

### Deployment

Docker-Compose

### Deployment details

Docker Compose version v2.0.0

### What happened

The webserver does not correctly start

```
mwaa-airflow-webserver-1  | [2021-10-21 14:52:39,932] {providers_manager.py:156} WARNING - Exception when importing 'airflow.providers.google.leveldb.hooks.leveldb.LevelDBHook' from 'apache-airflow-providers-google' package: No module named 'plyvel'
mwaa-airflow-webserver-1  | [2021-10-21 14:52:40,842] {providers_manager.py:156} WARNING - Exception when importing 'airflow.providers.google.leveldb.hooks.leveldb.LevelDBHook' from 'apache-airflow-providers-google' package: No module named 'plyvel'
mwaa-airflow-webserver-1  |   ____________       _____________
mwaa-airflow-webserver-1  |  ____    |__( )_________  __/__  /________      __
mwaa-airflow-webserver-1  | ____  /| |_  /__  ___/_  /_ __  /_  __ \_ | /| / /
mwaa-airflow-webserver-1  | ___  ___ |  / _  /   _  __/ _  / / /_/ /_ |/ |/ /
mwaa-airflow-webserver-1  |  _/_/  |_/_/  /_/    /_/    /_/  \____/____/|__/
mwaa-airflow-webserver-1  | [2021-10-21 14:52:40,903] {dagbag.py:500} INFO - Filling up the DagBag from /dev/null
mwaa-airflow-webserver-1  | [2021-10-21 14:52:41,525] {manager.py:512} WARNING - Refused to delete permission view, assoc with role exists View Menus.can_create Admin
mwaa-airflow-webserver-1  | [2021-10-21 14:52:41,538] {manager.py:512} WARNING - Refused to delete permission view, assoc with role exists View Menus.can_edit Admin
mwaa-airflow-webserver-1  | [2021-10-21 14:52:41,557] {manager.py:512} WARNING - Refused to delete permission view, assoc with role exists View Menus.can_delete Admin
mwaa-airflow-webserver-1  | [2021-10-21 14:52:41,886] {manager.py:512} WARNING - Refused to delete permission view, assoc with role exists DAG Runs.can_create User
mwaa-airflow-webserver-1  | [2021-10-21 14:52:41,901] {manager.py:512} WARNING - Refused to delete permission view, assoc with role exists DAG Runs.clear Admin
mwaa-airflow-webserver-1  | usage: airflow [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
mwaa-airflow-webserver-1  |    or: airflow --help [cmd1 cmd2 ...]
mwaa-airflow-webserver-1  |    or: airflow --help-commands
mwaa-airflow-webserver-1  |    or: airflow cmd --help
mwaa-airflow-webserver-1  | 
mwaa-airflow-webserver-1  | error: invalid command 'webserver'
mwaa-airflow-webserver-1 exited with code 1
```

### How to reproduce

`docker-compose up airflow-init`
`docker-compose up`

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by mik-laj** [comment]
What's your docker-compose.yaml file? What Docker image are you using?

**Comment by mik-laj** [comment]
```
mwaa-airflow-webserver-1
```
If you are having trouble running the latest Airflow image in [aws-mwaa-local-runner](https://github.com/aws/aws-mwaa-local-runner) then you should contact AWS Technical Support.

**Comment by DanielMorales9** [comment]
I am using the latest docker-compose from the documentation page. This has nothing to do with AWS.

**Comment by potiuk** [comment]
So why `mwaa-airflow-webserver-1` name ? Looks like a strange name to keep the "plain" airflow image in.

More quesitons:

* Did you follow precistly the instructions from https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html?
* Did you modify any of the  configuration/files other than the steps specified in the instructions ?
* Did you use official airflow image to run it (or did you use the original image) - which version? 
* Did you modify image in any way or used a different one? 
* Could you please copy &paste the whole output showing the exact image used ?

I think there is no easy reproduction without answers and likely the answers will indicate that there were some modifcation you've done. Conveting that into discussion until more information is provided.

**Reply by DanielMorales9** [reply]
Hi @potiuk, I am very new to Airflow and I am setting up a local development environment using the docker-compose the document provides. The idea is to develop locally my dags and then deploy them to MWAA,
[logs.txt](https://github.com/apache/airflow/files/7393329/logs.txt)
 however, I am not there yet.

Coming to your questions:

1. I followed precisely the instructions from the documentation
2. I did not modify any of the conf/files
3. I used the official apache airflow image `apache/airflow:2.2.0`
4. I did not modify the image in any way or use a different one
5. I attached the output of the docker-compose

**Reply by potiuk** [reply]
Are you sure you did the steps here ?  https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html#initializing-environment

The logs indicate at least part of those has not been done or maybe you have some more serious problem.wirh your docker environment . 

What is the environment /OS docker-compose versions you have ?

Are you running in on windows?

**Reply by potiuk** [reply]
And what is your .env file content ?

**Reply by mik-laj** [reply]
Can you try to use install Airflow 2.1.4? Airflow 2.2.0 have known bugs and may cause problems with Docker Compose.  Until now, I have not been able to prepare an environment that would fully boot.

**Reply by mik-laj** [reply]
Here guide and docker compose: https://airflow.apache.org/docs/apache-airflow/2.1.4/start/docker.html

**Reply by DanielMorales9** [reply]
@potiuk I have Mac OS, docker-compose v2.0. 
It does not make any difference if you have set the AIRFLOW_PID or not. 
I noticed, though, that when your dags or plugins folder is non-empty it does not work. This is the only relation that I found between working runs and non working runs.

**Reply by DanielMorales9** [reply]
@mik-laj that must be a complete different docker-compose file right? I will try it and let you know!

**Reply by mik-laj** [reply]
Yes. Docker compose file for Airflow 2.1 have a one more service and different image.

**Reply by Nidesh-Arumugam** [reply]
i have already used airflow 2.1.3 using docker , it's still working fine for me. But they said  patch endpoint  to update the dag_run state was added in 2.2. Any idea, why server is not running in 2.2?

**Reply by DanielMorales9** [reply]
Same situation, when dags and plugins are non-empty it does not start

**Reply by DanielMorales9** [reply]
My plugins directory seems to disturb the deployment process

**Reply by potiuk** [reply]
Yeah. If you have custom plugin, then it might destroy things.

 I'd say one of your plugins have Sys.exit() call somewhere which is not generating a visible error or warning or exception but causes webserver to exit

**Reply by potiuk** [reply]
What you can do is to remove all plugins and add one by one and see which one is causing the problem.

**Reply by DanielMorales9** [reply]
Ok, I can confirm it was my plugins directory. It must have been the setup.py or some other conf file. 
Do I need to file an issue to validate the directory structure? Maybe I can follow up with a PR.

**Reply by potiuk** [reply]
I think it is indeed something you've done in your setup - either you simply had sys.exit(0) somewhere in your plugin top code - which made it fail without error or warning or (more likely) your code somewhat overrode the default 'airflow' command entrypoint and what 'airflow webserver' executed was your entrypoint from your setup py rather than airflow one.

I vaguely recall similar problem from the past but it would be great to precisely pin-point the issue before we decide if we can do something about it or just add a documentation.

 I seriously doubt you can do any validation. Of such or similar mistake. Python is flexible language if you 'exit' interpreter, it will exit. There is not much you can do there. I am not sure what exactly was the content of your setup.py in your plugin but it would be great if you share it here (if you can) and maybe we can find out together what is the remediation.

If we can find exactly what you did there, it might be great to at least add a note (and mention error you get in this case) about it in our best practices in Modules Management chapter even if we can't automatically detect it and provide better error message.

We already have a number of DO's and DONT's: they explain how your folders should be structured.
 https://airflow.apache.org/docs/apache-airflow/stable/modules_management.html

**Comment by Nidesh-Arumugam** [comment]
found a way Daniel ? Even i am facing the same issue.

**Reply by mdcsaenz** [reply]
Im also facing this issue with apache/airflow:2.2.2

**Reply by Nidesh-Arumugam** [reply]
I got the issue solved. For me, i got it solved  by increasing the allocated  ram size for docker  4 to 8gb. Then the airflow server started as expected.
